src/main/java/com/opentable/kafka/util/OffsetMetrics.java

```diff
@@ -234,21 +269,85 @@ private void pollUnsafe(final String topic) {
+        unknownParts.forEach(              part        -> gauge(partitionName(topic, part, "size")).set(null));
+        gauge(totalName(topic, "size")).set(sumValues(subMap(knownParts, sizes)));
 
+        offsets     .forEach((part, off) -> gauge(partitionName(topic, part, "offset")).set(off));
+        unknownParts.forEach( part       -> gauge(partitionName(topic, part, "offset")).set(null));
         gauge(totalName(topic, "offset")).set(sumValues(offsets));
-        gauge(totalName(topic, "lag")).set(sumValues(lag));
+
+        lags.forEach((part, lag) -> {
+            gauge(partitionName(topic, part, "lag")).set(lag);
+            ((Histogram) metricMap.get(totalName(topic, "lag-distribution"))).update(lag);
+        });
+        unknownParts.forEach(part -> gauge(partitionName(topic, part, "lag")).set(null));
+        gauge(totalName(topic, "lag")).set(sumValues(lags));
     }
 
-    private AtomicLongGauge gauge(final String name) {
-        return (AtomicLongGauge) metricMap.get(name);
+    private LongGauge gauge(final String name) {
+        return (LongGauge) metricMap.get(name);
     }
 
     private static long sumValues(final Map<?, Long> map) {
-        return map.values().stream().mapToLong(size -> size).sum();
+        return map.values().stream().mapToLong(x -> x).sum();
+    }
+
+    /**
+     * Sub-map of {@code map} whose keys are the elements of {@code keys} with corresponding values from {@code map}.
+     */
+    private static <K, V> Map<K, V> subMap(final Collection<K> keys, final Map<K, V> map) {
+        return new AbstractMap<K, V>() {
+            @Override
+            public Set<Entry<K, V>> entrySet() {
+                return new AbstractSet<Entry<K, V>>() {
+                    @Override
+                    public Iterator<Entry<K, V>> iterator() {
+                        final Iterator<Entry<K, V>> itr = map.entrySet().iterator();
+                        return new AbstractIterator<Entry<K, V>>() {
+                            @Override
+                            protected Entry<K, V> computeNext() {
+                                while (itr.hasNext()) {
+                                    final Entry<K, V> e = itr.next();
+                                    if (keys.contains(e.getKey())) {
+                                        return e;
+                                    }
+                                }
+                                return endOfData();
+                            }
+                        };
+                    }
+
+                    @Override
+                    public int size() {
+                        return keys.size();
+                    }
+                };
+            }
+        };
+    }
+
+    @VisibleForTesting
+    static class LongGauge implements Gauge<Long>, Counting {
+        private final AtomicReference<Long> value = new AtomicReference<>();
```

Comment:
You could hold an `AtomicLong` instead of `AtomicReference<Long>` or if you really want to you could even `extends AtomicLong`

Reply:
Or moreover, use our `AtomicLongGauge`?

No to all three.  `AtomicLong`s can't be `null`.